pkg/engine/executor/parse_json.go

```diff
@@ -0,0 +1,244 @@
+		j.prefixBuffer = append(j.prefixBuffer, key)
+
+		// Recursively parse nested object
+		err := jsonparser.ObjectEach(value, func(nestedKey []byte, nestedValue []byte, nestedType jsonparser.ValueType, _ int) error {
+			return j.parseObject(nestedKey, nestedValue, nestedType, result, requestedKeyLookup)
+		})
+
+		// rollback the prefix as we exit the current object
+		j.prefixBuffer = j.prefixBuffer[:prefixLen]
+		return err
+	case jsonparser.Array:
+		// Skip arrays for now (same behavior as original)
+		return nil
+	case jsonparser.Null:
+		// Skip null values
+		return nil
+	default:
+		// Skip unknown types
+		return nil
+	}
+}
+
+func (j *JSONParser) parseLabelValue(key, value []byte, dataType jsonparser.ValueType, result map[string]string, requestedKeyLookup map[string]struct{}) error {
+	// Build the full key (with flattening for nested objects)
+	var keyString string
+	if len(j.prefixBuffer) == 0 {
+		// Top-level key case
+		// key is a slice pointing to the same underlying array as the data passed to jsonparser.ObjectEach
+		// nothing should mutate this data as it's the original JSON string, so it should be safe to use unsafeString
+		keyString = sanitizeLabelKey(unsafeString(key), true)
+	} else {
+		// Nested key - build flattened key with underscore separator
+		j.prefixBuffer = append(j.prefixBuffer, key)
+		sanitized := j.buildSanitizedPrefixFromBuffer()
+		// Make a copy to avoid buffer reuse issues, since sanitized is a slice using the same array behind j.sanitizedPrefixBuffer
+		keyString = string(sanitized)
+		// Remove the key we just added since we'll add it back in parseObject
+		j.prefixBuffer = j.prefixBuffer[:len(j.prefixBuffer)-1]
+	}
+
+	// Filter by requested keys if provided
+	if requestedKeyLookup != nil {
+		if _, wantKey := requestedKeyLookup[keyString]; !wantKey {
+			return nil // Skip this key
+		}
+	}
+
+	// Convert the value to string based on its type
+	j.valueBuffer = readValue(value, dataType, j.valueBuffer)
+	if j.valueBuffer != nil {
+		result[keyString] = string(j.valueBuffer) // make a copy as j.valueBuffer is reused
+	}
+
+	return nil
+}
+
+func (j *JSONParser) buildSanitizedPrefixFromBuffer() []byte {
+	j.sanitizedPrefixBuffer = j.sanitizedPrefixBuffer[:0]
+
+	for i, part := range j.prefixBuffer {
+		if len(bytes.TrimSpace(part)) == 0 {
+			continue
+		}
+
+		if i > 0 && len(j.sanitizedPrefixBuffer) > 0 {
+			j.sanitizedPrefixBuffer = append(j.sanitizedPrefixBuffer, byte(jsonSpacer))
+		}
+		j.sanitizedPrefixBuffer = appendSanitized(j.sanitizedPrefixBuffer, part)
+	}
+
+	return j.sanitizedPrefixBuffer
+}
+
+func readValue(v []byte, dataType jsonparser.ValueType, buf []byte) []byte {
+	switch dataType {
+	case jsonparser.String:
+		return unescapeJSONString(v, buf)
+	case jsonparser.Null:
+		return nil
+	case jsonparser.Number:
+		return v
+	case jsonparser.Boolean:
+		if bytes.Equal(v, trueBytes) {
+			return trueBytes
+		}
+		return falseBytes
+	default:
+		return nil
+	}
```

Comment:
There's an obscure corruption bug here where if you parse an escaped string after parsing a boolean, you'll corrupt the global value for trueBytes/falseBytes. We should copy this code from an older version of main that doesn't have #18637 in it, which introduced that corruption bug.

Reply:
wow, really good catch! I remember seeing that bug being discussed and totally missed it was still here in my copy pasta.